src/main/java/nextstep/security/authentication/DaoAuthenticationProvider.java

```diff
@@ -0,0 +1,32 @@
+package nextstep.security.authentication;
+
+import nextstep.app.ui.AuthenticationException;
+import nextstep.security.UserDetails;
+import nextstep.security.UserDetailsService;
+
+import java.util.Objects;
+
+public class DaoAuthenticationProvider implements AuthenticationProvider {
+
+    private final UserDetailsService userDetailsService;
+
+    public DaoAuthenticationProvider(final UserDetailsService userDetailsService) {
+        this.userDetailsService = userDetailsService;
+    }
+
+    @Override
+    public Authentication authenticate(final Authentication authentication) throws AuthenticationException {
+        final UserDetails userDetails = userDetailsService.loadUserByUsername((String) authentication.getPrincipal());
+
+        if (!Objects.equals(userDetails.getPassword(), authentication.getCredentials())) {
+            throw new AuthenticationException();
+        }
+
+        return new UsernamePasswordAuthenticationToken(userDetails.getPassword(), userDetails.getUsername(), true);
```

Comment:
UsernamePasswordAuthenticationToken을 생성하는 부분에서 인자 순서가 잘못되었네요 😢  
```suggestion
        return new UsernamePasswordAuthenticationToken(userDetails.getUsername(), userDetails.getPassword(), true);
```